### PR TITLE
add new ENV node in addons.yaml

### PR DESCRIPTION
**Pull request by yelizariev**
closes #258

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -343,6 +346,14 @@ ONLY:
     - prod
 server-tools:
   - auth_*
+---
+# Custom repositories
+ENV:
+  DEFAULT_REPO_PATTERN: https://github.com/it-projects-llc/{}.git
```
Comment (pedrobaeza) [inline README.md]: I would prefer here to not point to other specific repo and let documentation "brand agnostic".

Comment (yelizariev) [reply README.md]: I changed it to Tecnativa, as it's already used in the README
https://github.com/Tecnativa/doodba/compare/1d563c5050a228bb97fa52459dfa7c8f7280760e..1a5a25f3e1123c4c2e52793ba45e79097fead486

Comment (pedrobaeza) [reply README.md]: Well, it's not needed to put `Tecnativa`. I was thinking in a generic `team` word, but thanks for branding us.

Comment (yelizariev) [reply README.md]: I decided that github may not have something like ``example.com`` and any generic ``team`` may be occupied later

Comment (pedrobaeza) [reply README.md]: OK, thanks

File: bin/autoaggregate [hunk 2]
```diff
@@ -77,7 +77,7 @@ def origin_for(folder):
         pattern.
     """
     base = os.path.basename(folder)
-    pattern = DEFAULT_REPO_PATTERN
+    pattern = env.get("DEFAULT_REPO_PATTERN", DEFAULT_REPO_PATTERN)
```
Comment (Yajo) [inline bin/autoaggregate]: Thanks for the test. I still don't like that this isn't actually setting environment variables as expected, but just using this one.

Please at least add a TODO here and a note in the README so this weird thing is clear.

Comment (pedrobaeza) [reply bin/autoaggregate]: Can't this iterate in the values found in this section and set that variables in the environment?

Comment (Yajo) [reply bin/autoaggregate]: That would be a better solution indeed.

We usually recommend using things such as `$DEPTH_DEFAULT` or `$ODOO_VERSION`. I don't remember if the autogenerated `repos.yaml` includes those values or expands them directly.

Respecting all env variables declared would be best indeed.

Comment (yelizariev) [reply bin/autoaggregate]: Possible updates are as follows:

1. _Use_ other environment variables (DEPTH_DEFAULT and ODOO_VERSION) on creating config: https://github.com/Tecnativa/doodba/blob/e40eb51500e3837767da440e16b3da63e3b41fb5/bin/autoaggregate#L122-L125

2. Create auto repos.yaml per environment (or per repository for simplicity) and call ``aggregate`` for each of them https://github.com/Tecnativa/doodba/blob/e40eb51500e3837767da440e16b3da63e3b41fb5/bin/autoaggregate#L138-L142

Comment (Yajo) [reply bin/autoaggregate]: Better option 1. We need to declare all repos in a single file to benefit from parallel downloading.

So, just expand all variables when generating the repos file.

**Comment by yelizariev** [comment]
All seems done

File: bin/autoaggregate [hunk 3]
```diff
@@ -98,31 +98,36 @@ def missing_repos_config():
                     defined.add(os.path.abspath(os.path.join(SRC_DIR, repo)))
     except (IOError, AttributeError):
         logger.debug("No repositories defined by hand")
+    addons_env = {}
     # Find the repositories that should be present
     try:
         with open(ADDONS_YAML) as yaml_file:
             for doc in yaml.safe_load_all(yaml_file):
+                env = doc.get("ENV", {})
```
Comment (Yajo) [inline bin/autoaggregate]:
```suggestion
                env = dict(os.environ, **doc.get("ENV", {}))
```

File: bin/autoaggregate [hunk 4]
```diff
@@ -98,31 +98,36 @@ def missing_repos_config():
...
-    config = {
-        repo_path: {
-            "defaults": {"depth": "$DEPTH_DEFAULT"},
-            "merges": ["origin $ODOO_VERSION"],
-            "remotes": {"origin": origin_for(repo_path)},
-            "target": "origin $ODOO_VERSION",
+    config = {}
+    for repo_path in missing:
+        env = addons_env.get(repo_path, {})
+        depth = env.get("DEPTH_DEFAULT", "$DEPTH_DEFAULT")
```
Comment (Yajo) [inline bin/autoaggregate]:
```suggestion
        depth = env["DEPTH_DEFAULT"]
```

File: bin/autoaggregate [hunk 5]
```diff
@@ -98,31 +98,36 @@ def missing_repos_config():
...
+        origin_version = "origin %s" % env.get("ODOO_VERSION", "$ODOO_VERSION")
```
Comment (Yajo) [inline bin/autoaggregate]:
```suggestion
        origin_version = "origin %s" % env["ODOO_VERSION"]
```

File: bin/autoaggregate [hunk 6]
```diff
@@ -68,7 +68,7 @@ def aggregate(config):
                         )
 
 
-def origin_for(folder):
+def origin_for(folder, env):
```
Comment (Yajo) [inline bin/autoaggregate]:
```suggestion
def origin_for(folder, default_repo_pattern=DEFAULT_REPO_PATTERN, odoo_repo_pattern=DEFAULT_REPO_PATTERN_ODOO):
```

File: bin/autoaggregate (hunk 2 above)
Comment (Yajo) [inline bin/autoaggregate]:
```suggestion
    pattern = default_repo_pattern
```

File: bin/autoaggregate [hunk 7]
```diff
@@ -77,7 +77,7 @@ def origin_for(folder):
         pattern.
     """
     base = os.path.basename(folder)
-    pattern = DEFAULT_REPO_PATTERN
+    pattern = env.get("DEFAULT_REPO_PATTERN", DEFAULT_REPO_PATTERN)
     if base == "odoo":
         pattern = DEFAULT_REPO_PATTERN_ODOO
```
Comment (Yajo) [inline bin/autoaggregate]:
```suggestion
         pattern = odoo_repo_pattern
```

File: bin/autoaggregate [hunk 8]
```diff
@@ -98,31 +98,36 @@ def missing_repos_config():
...
-            "target": "origin $ODOO_VERSION",
+    config = {}
+    for repo_path in missing:
+        env = addons_env.get(repo_path, {})
+        depth = env.get("DEPTH_DEFAULT", "$DEPTH_DEFAULT")
+        origin_version = "origin %s" % env.get("ODOO_VERSION", "$ODOO_VERSION")
+        config[repo_path] = {
+            "defaults": {"depth": depth},
+            "merges": [origin_version],
+            "remotes": {"origin": origin_for(repo_path, env)},
```
Comment (Yajo) [inline bin/autoaggregate]:
```suggestion
            "remotes": {"origin": origin_for(repo_path, env["DEFAULT_REPO_PATTERN"], env["DEFAULT_REPO_PATTERN_ODOO"])},
```

File: README.md [hunk 9]
```diff
@@ -297,6 +297,13 @@ Advanced features:
 - You can use `ONLY` to supply a dictionary of environment variables and a
   list of possible values to enable that document in the matching environments.
 
+- You can use `ENV` to supply a dictionary of environment variables to be used
+  on downloading repositories. Following variables are supported:
+
+  - `DEFAULT_REPO_PATTERN`
```
Comment (Yajo) [inline README.md]:
```suggestion
  - `DEFAULT_REPO_PATTERN`
  - `DEFAULT_REPO_PATTERN_ODOO`
```

File: tests/__init__.py [hunk 10]
```diff
@@ -270,6 +270,18 @@ def test_smallest(self):
...
+    def test_addons_env(self):
+        """Test custom pattern in addons.yaml"""
+        # 7.0 is skiped because the module is absent in that branch
+        for sub_env in matrix(odoo_skip={"7.0"}):
+            self.compose_test(
+                join(SCAFFOLDINGS_DIR, "addons_env"),
+                sub_env,
+                ("test", "-d", "custom/src/misc-addons"),
+                ("test", "-d", "custom/src/misc-addons/web_debranding"),
+                ("test", "-e", "auto/addons/web_debranding"),
```
Comment (Yajo) [inline tests/__init__.py]: Please also test support for `DEFAULT_REPO_PATTERN_ODOO`.

File: bin/autoaggregate [hunk 11]
```diff
@@ -98,31 +103,45 @@ def missing_repos_config():
...
         with open(ADDONS_YAML) as yaml_file:
             for doc in yaml.safe_load_all(yaml_file):
+                env = dict(os.environ, **doc.get("ENV", {}))
                 for repo in doc:
-                    if repo in {PRIVATE, CORE, "ONLY"}:
+                    if repo in {PRIVATE, "ONLY", "ENV"}:
                         continue
-                    repo_path = os.path.abspath(os.path.join(SRC_DIR, repo))
+                    if repo == CORE:
+                        repo_path = repo_path_odoo
```
Comment (yelizariev) [inline bin/autoaggregate]: Some addition refactoring

File: README.md [hunk 12]
```diff
@@ -249,14 +249,16 @@ download git code:
 
 ```yaml
 /opt/odoo/custom/src/odoo:
-  depth: $DEPTH_DEFAULT
+  defaults:
+    depth: $DEPTH_DEFAULT
```
Comment (yelizariev) [inline README.md]: Fixes in readme

File: tests/__init__.py [hunk 13]
```diff
@@ -270,6 +270,22 @@ def test_smallest(self):
...
+            self.compose_test(
+                join(SCAFFOLDINGS_DIR, "addons_env"),
+                sub_env,
+                # check module from custom repo pattern
+                ("test", "-d", "custom/src/misc-addons"),
+                ("test", "-d", "custom/src/misc-addons/web_debranding"),
+                ("test", "-e", "auto/addons/web_debranding"),
+                # Migrations folder is only in OpenUpgrade
+                ("test", "-e", "auto/addons/crm"),
+                ("test", "-d", "auto/addons/crm/migrations"),
```
Comment (yelizariev) [inline tests/__init__.py]: Tests for ``DEFAULT_REPO_PATTERN_ODOO``

File: tests/scaffoldings/addons_env/custom/src/addons.yaml [hunk 14]
```diff
@@ -0,0 +1,14 @@
+server-tools:
+  - module_auto_update
+---
+ENV:
+  DEFAULT_REPO_PATTERN_ODOO: https://github.com/OCA/OpenUpgrade.git
+odoo/addons:
+  - crm
+  - sale
```
Comment (yelizariev) [inline tests/scaffoldings/addons_env/custom/src/addons.yaml]: test config for odoo

File: bin/autoaggregate [hunk 15]
```diff
@@ -98,31 +103,45 @@ def missing_repos_config():
...
     missing = expected - defined
-    config = {
-        repo_path: {
-            "defaults": {"depth": "$DEPTH_DEFAULT"},
-            "merges": ["origin $ODOO_VERSION"],
-            "remotes": {"origin": origin_for(repo_path)},
-            "target": "origin $ODOO_VERSION",
+    config = {}
+    for repo_path in missing:
+        env = addons_env.get(repo_path, os.environ)
```
Comment (yelizariev) [inline bin/autoaggregate]: ``os.environ`` is default env

File: bin/autoaggregate [hunk 16]
```diff
@@ -89,7 +93,8 @@ def missing_repos_config():
     :return dict:
         git-aggregator-ready configuration dict for undefined repositories.
     """
-    defined, expected = set(), {os.path.join(SRC_DIR, "odoo")}
+    repo_path_odoo = os.path.join(SRC_DIR, "odoo")
+    defined, expected = set(), {repo_path_odoo}
```
Comment (Yajo) [inline bin/autoaggregate]: You should add to the top `from doodbalib import ODOO_DIR`. Then:

```suggestion
    defined, expected = set(), {ODOO_DIR}
```

**Comment by yelizariev** [comment]
Yeah, I figure it out from ``travis.yml``. I just forget to run it

> I noticed some linter failures, you can simply run pre-commit install and it will make sure all is properly formatted before pushing. wink

**Pull request merged**